torch/csrc/jit/script/compiler.cpp

```diff
@@ -472,7 +472,11 @@ Value* tryMatchArgument(
   if (value->type()->isSubtypeOf(NoneType::get())){
     if (concrete_type->isSubtypeOf(GeneratorType::get())) {
       value = graph.insertNode(graph.createNoneGenerator())->output();
-    } else if (concrete_type->isSubtypeOf(DynamicType::get())) {
+    }
+    auto concrete_opt_type = concrete_type->cast<OptionalType>();
+    if ((concrete_opt_type && concrete_opt_type->getElementType()->isSubtypeOf(DynamicType::get())) ||
+        concrete_type->isSubtypeOf(DynamicType::get())) {
```

Comment:
We should no longer allow None to be passed to a Tensor. We should only allow it to be passed to a Tensor?
This might be more cleanly written as:
```
concrete_type->isSubclassOf(OptionType::ofTensor())
```
See `ListType::ofTensors()` as an example of the idiom.

Reply:
Yeah I thought about this, I leave it there because there were several places in our nn test that are accepting a None type and are not converted to weak script yet. I will see if I could get all these works

Reply:
Unfortunately, a method cannot be both `virtual` and `static`, there might be a solution if we want to go towards the static `isSubtypeOf` approach as indicated [here](https://stackoverflow.com/questions/34222703/how-to-override-static-method-of-template-class-in-derived-class)